dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,316 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <pch.h>
+#include <future>
+#include <PowerNotifications.h>
+#include <frameworkudk\PowerNotificationsPal.h>
+#include <PowerManager.g.cpp>
+#include <powrprof.h>
+
+namespace winrt::Microsoft::ProjectReunion::implementation
+{
+
+    // EnergySaverStatus Functions
+    EventType& EnergySaverStatus_Event()
+    {
+        return make_self<factory_implementation::PowerManager>()->m_energySaverStatusChangedEvent;
+    }
+
+    void EnergySaverStatus_Register()
+    {
+        check_hresult(PowerNotifications_RegisterEnergySaverStatusChangedListener(
+            &PowerManager::EnergySaverStatusChanged_Callback,
+            &make_self<factory_implementation::PowerManager>()->m_energySaverStatusHandle));
+    }
+
+    void EnergySaverStatus_Unregister()
+    {
+        check_hresult(PowerNotifications_UnregisterEnergySaverStatusChangedListener(
+            make_self<factory_implementation::PowerManager>()->m_energySaverStatusHandle));
+    }
+
+    void EnergySaverStatus_Update()
+    {
+        check_hresult(PowerNotifications_GetEnergySaverStatus(
+            &make_self<factory_implementation::PowerManager>()->m_cachedEnergySaverStatus));
+    }
+
+    // BatteryStatus Functions
+    EventType& BatteryStatus_Event()
+    {
+        return make_self<factory_implementation::PowerManager>()->m_batteryStatusChangedEvent;
+    }
+
+    void BatteryStatus_Register()
+    {
+        check_hresult(PowerNotifications_RegisterCompositeBatteryStatusChangedListener(
+            &PowerManager::CompositeBatteryStatusChanged_Callback,
+            &make_self<factory_implementation::PowerManager>()->m_batteryStatusHandle));
+    }
+
+    void BatteryStatus_Unregister()
+    {
+        check_hresult(PowerNotifications_UnregisterCompositeBatteryStatusChangedListener(
+            make_self<factory_implementation::PowerManager>()->m_batteryStatusHandle));
+    }
+
+    void BatteryStatus_Update()
+    {
+        auto status = &make_self<factory_implementation::PowerManager>()->m_cachedCompositeBatteryStatus;
+        if (PowerNotifications_GetCompositeBatteryStatus(status) == S_OK)
```

Comment:
No update if there is a failure?  Silently let it pass?

Reply:
Yes, this is in cases where there is no battery on the system.

Reply:
Does `PowerNotifications_GetCompositeBatteryStatus()` return any SUCCEEDED() HRESULT other than S_OK?

Some old APIs can (e.g. MSXML.loadXML() can return S_OK or S_FALSE as different flavors of success), but that's been out of vogue for a long time now. As long as `PowerNotifications_GetCompositeBatteryStatus()` doesn't return varying values you need to distinguish you should do `if (SUCCEEDED(PowerNotifications_GetCompositeBatteryStatus(status)))`

(Any code explicitly checking ==/!=S_OK immediately flags suspicion as possibly flawed)

Reply:
Changed

Reply:
This is still being compared to S_OK

Reply:
Github is confusing; you are referring to old code.

And if there was a battery a filled compositebattery struct would be returned and we would calculate the rest based on the values in it.